### PR TITLE
feat: Checkpointing for canister snapshots

### PR DESCRIPTION
**Pull request by schneiderstefan**
This commit implements writing canister snapshots to disk as part of checkpointing and loading them back.

### CODE REVIEW

File: rs/state_manager/src/lib.rs [hunk 1]
```diff
@@ -1139,6 +1171,13 @@ impl PageMapType {
             PageMapType::WasmChunkStore(id) => state
                 .canister_state_mut(id)
                 .map(|can| can.system_state.wasm_chunk_store.page_map_mut()),
+            // Snapshots are kept behind an `Arc` inside `state`, which makes the pattern of `list_all` and `get_mut` for pagemaps difficult
+            // without excessive copies.
+            // In the few places where we need to iterate over all pagemaps and modify them we instead handle snapshots separately.
+            // Note that this is not a problem in the many more places where we need to iterate over all pagemaps and only need read access.
+            PageMapType::SnapshotWasmMemory(_) => unimplemented!(),
+            PageMapType::SnapshotStableMemory(_) => unimplemented!(),
+            PageMapType::SnapshotWasmChunkStore(_) => unimplemented!(),
```
Comment (pakhomov-dfinity) [inline rs/state_manager/src/lib.rs:1180]: I wonder whether splitting the PageMapType into PageMapType aas here plus SnapshotPageMApType would make it easier, as snapshots need a different treatment and can be a different type to make it explicit. But not sure that could work.

**Review by dsarlis — APPROVED** [review]
Approving execution changes.

File: rs/state_manager/src/tip.rs [hunk 2]
```diff
@@ -436,10 +450,11 @@ pub(crate) fn spawn_tip_thread(
                             execution_state,
                         } => {
                             debug_assert!(
-                                checkpointed_state == execution_state,
+                                checkpointed_state.canister_snapshots
+                                    == execution_state.canister_snapshots,
```
Comment (pakhomov-dfinity) [inline rs/state_manager/src/tip.rs:454]: Leftover

File: rs/state_manager/tests/state_manager.rs [hunk 3]
```diff
@@ -5922,6 +5925,210 @@ fn lsmt_shard_size_is_stable() {
...
+            };
+
+            verify_state(&state);
+            state_manager.commit_and_certify(state, height(4), CertificationScope::Full, None);
+            let (_height, state) = state_manager.take_tip();
+            verify_state(&state);
+            state_manager.commit_and_certify(state, height(5), CertificationScope::Full, None);
+            let (_height, state) = state_manager.take_tip();
+            verify_state(&state);
+            state_manager.commit_and_certify(state, height(6), CertificationScope::Full, None);
```
Comment (pakhomov-dfinity) [inline rs/state_manager/tests/state_manager.rs:6125]: Why exactly 3 checkpoints? Should it be a loop?

File: rs/replicated_state/src/canister_state/system_state/wasm_chunk_store.rs [hunk 4]
```diff
@@ -167,7 +167,7 @@ impl WasmChunkStore {
         Ok(hash)
     }
 
-    pub(crate) fn from_checkpoint(data: PageMap, metadata: WasmChunkStoreMetadata) -> Self {
+    pub fn from_checkpoint(data: PageMap, metadata: WasmChunkStoreMetadata) -> Self {
```
Comment (pakhomov-dfinity) [inline rs/replicated_state/src/canister_state/system_state/wasm_chunk_store.rs:170]: How did we use to handle it for non-snapshot WasmChunkStore?

Comment (schneiderstefan) [reply rs/replicated_state/src/canister_state/system_state/wasm_chunk_store.rs:170]: This function is called from the constructor of `SystemState`, which is in the same crate. Hence the previous `pub(crate)`.

File: rs/state_layout/src/state_layout.rs [hunk 5]
```diff
@@ -1565,6 +1565,50 @@ impl<Permissions: AccessPolicy> PageMapLayout<Permissions> {
 
         Ok(result)
     }
+
+    /// Helper function to copy the files from `PageMapsLayout` `src` to another `PageMapLayout` `dst`.
+    /// This is used in the context of canister snapshots, where files need to be copied from a canister
+    /// to a snaphsot or vice versa.
+    pub fn copy_or_hardlink_files<W>(
```
[the remark this answers is not in the document]
Comment (schneiderstefan) [reply rs/state_layout/src/state_layout.rs:1572]: It makes fewer assumptions. Right now I am saying "copy all pagemaps and copy the binary". If the code says "copy everything" I have to rely on the fact that the the protobufs (or anything else for that matter) aren't in the canister's tip directory at this point in time.

File: rs/state_manager/src/lib.rs (hunk 1 above)
Comment (schneiderstefan) [reply rs/state_manager/src/lib.rs:1180]: Yeah I am not particularly happy with this bit either. I was trying out a few things but didn't really find anything that worked.

I am very open to revisiting this part of the code again, but if it's ok I wouldn't do that as part of this PR.

**Review by pakhomov-dfinity — APPROVED** [review]

**Pull request merged**